src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/create/HaFlowCreateFsm.java

```diff
@@ -0,0 +1,432 @@
+    private boolean pathsBeenAllocated;
+    private boolean backUpPrimaryPathComputationWayUsed;
+    private boolean backUpProtectedPathComputationWayUsed;
+    private Map<PathId, Boolean> backUpComputationWayUsedMap;
+    private final List<SpeakerData> sentCommands;
+
+    // The amount of flow create operation retries left: that means how many retries may be executed.
+    // NB: it differs from command execution retries amount.
+    private int remainRetries;
+    private boolean timedOut;
+
+    private HaFlowCreateFsm(@NonNull CommandContext commandContext, @NonNull FlowGenericCarrier carrier,
+                            @NonNull String haFlowId,
+                            @NonNull Collection<FlowCreateEventListener> eventListeners, @NonNull Config config) {
+        super(Event.NEXT, Event.ERROR, commandContext, carrier, haFlowId, eventListeners);
+        this.remainRetries = config.getHaFlowCreationRetriesLimit();
+        backUpComputationWayUsedMap = new HashMap<>();
+        sentCommands = new ArrayList<>();
+        yPointMap = new HashMap<>();
+    }
+
+    /**
+     * Initiates a retry if limit is not exceeded.
+     *
+     * @return true if retry was triggered.
+     */
+    public boolean retryIfAllowed() {
+        if (!timedOut && remainRetries-- > 0) {
+            log.info("About to retry flow create. Retries left: {}", remainRetries);
+            resetState();
+            fire(Event.RETRY);
+            return true;
+        } else {
+            if (timedOut) {
+                log.warn("Failed to create flow: operation timed out");
+            } else {
+                log.debug("Retry of flow creation is not possible: limit is exceeded");
+            }
+            return false;
+        }
+    }
+
+    public List<HaFlowResources> getHaFlowResources() {
+        List<HaFlowResources> result = new ArrayList<>();
+        for (HaFlowResources resources : new HaFlowResources[]{primaryResources, protectedResources}) {
+            if (resources != null) {
+                result.add(resources);
+            }
+        }
+        return result;
+    }
+
+    @Override
+    protected void afterTransitionCausedException(State fromState, State toState, Event event,
+                                                  HaFlowCreateContext context) {
+        super.afterTransitionCausedException(fromState, toState, event, context);
+        String errorMessage = getLastException().getMessage();
+        if (fromState == State.INITIALIZED || fromState == State.FLOW_VALIDATED) {
+            ErrorData error = new ErrorData(ErrorType.INTERNAL_ERROR, "Could not create flow",
+                    errorMessage);
+            Message message = new ErrorMessage(error, getCommandContext().getCreateTime(),
+                    getCommandContext().getCorrelationId());
+            sendNorthboundResponse(message);
+        }
+
+        fireError(errorMessage);
+    }
+
+    private void resetState() {
```

Comment:
Right, so it's like there are two states, or groups of states. One kind of state is related to resources and another kind of state related to retries. I understand it as we have two fsms here. One handles retries and another all those states defined below. 
I mean reset state doesn't fully reset the _state_ since it remembers the number of retries. This reset state is more like "prepare for an Entry event". That was confusing for me. But changes could be even more confusing. Maybe add some comment here or at declaration that it is intended and they are not part of "state" in that sense of state, or that they are not supposed to be changed in reset state.